source-code/core/src/lint/RFC.md

```diff
@@ -0,0 +1,19 @@
+# Proposal for the `lint` functionality
+
+## Types
+
+1. I have investigated if [this](https://github.com/orgs/inlang/discussions/319#discussioncomment-4850815) is possible and I could not make it work.
+
+	I have tried 3 different ways to do it. Take a look at the `experiment-1.ts` file to see the examples. I find variant 3 to be the best because it does not hide the properties behind an abstraction (see screenshots below). It can be easily extended in the future with additional properties.
+
+2. We need to extend the Resource type on each level (`Resource`, `Message`, `Pattern`) because each level could have lint information attached to it.
+
+	- "missing key" would be on the `Resource` level
+	- "missing parameter" would be on the `Message` level
+	- "unknown parameter" would be on the `Pattern` level
+
+	Maybe `Pattern` requires another level in the future, but if we solve the problem for the first 3 levels, adding the 4th should be easy.
+
+	Kind of similar to the `metadata` property which can be present on all levels. If we solve the problem for `lint`, we can get rid of the non-typed `metadata` property and plugin authors can use a slightly different approach (additional properties should be scoped to a "namespace"-like object). The result would be a `Resource` with correctly typed `metadata` property.
+
+	See `experiment-2.ts` for an example how this could be done. The only downside is, that those generics make the code harder to read. And VS Code will show jibberish when hovering over a variable (see screenshots below). But the only alternative I can see is to define each Permutation manually which is even worse.
```

Comment:
![image](https://user-images.githubusercontent.com/21335119/216772398-805e2f16-4b42-4c35-91f4-1e6a336d9ee7.png)

Reply:
Let's think about our goal with extensions and whether this RFC needs to come up with a "proper" API to extend types, or whether unions are sufficient. 

A quick and simple solution could be to extend AST nodes with a union just as we do with types until a better API is required. 

```ts
/**
 * This alternative uses type unions.
 */

import type * as ast from "../ast/index.js";

// ------------------------------------------------------------------------------------------------
// ast extensions

type Resource = ast.Resource & {
  body: Array<Message>;
  lint: {
    errors: Array<LintResult>;
    warnings: Array<LintResult>;
    hasErrors: boolean;
    hasWarnings: boolean;
  };
};

type Message = ast.Message & {
  lint: {
    errors: Array<LintResult>;
    warnings: Array<LintResult>;
  };
};

// ------------------------------------------------------------------------------------------------
// lint types

export type LintResult = {
  id: string;
  type: "warning" | "error";
  message: string;
};

// ------------------------------------------------------------------------------------------------
/// tests

function lint(resource: ast.Resource): Resource {
  return resource as any;
}

const resource = lint({} as any);

resource.lint.errors;

resource.body[0].lint.errors;

```

Reply:
This will not work if you have a second extension function like `cache`

![image](https://user-images.githubusercontent.com/21335119/216808385-14c09ead-77e7-4a03-b5ce-f7e4dd3b7fdc.png)

Reply:
Yep. Ran into the same issue.  

<img width="667" alt="CleanShot 2023-02-05 at 17 04 16@2x" src="https://user-images.githubusercontent.com/35429197/216830455-afc62d60-d3e2-4d0e-af6e-a6e54f79f0f7.png">

Maybe someone on the [TypeScript Discord](https://discord.com/invite/typescript) has feedback on experiment 2 / making the AST nodes generic?

Reply:
`withCache` would most likely be a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) object btw that does not "extend" the nodes. The proxy intercepts queries of @inlang/core/query. 

Hmmm that makes me wonder whether using proxies could be of benefit for `lint` too

Reply:
Probably not. the AST should be serializable + using proxies to extend the AST ... why? The simple solution should just be to return an extended AST node. The question is just how we nail the types.

Reply:
> Maybe someone on the [TypeScript Discord](https://discord.com/invite/typescript) has feedback on experiment 2 / making the AST nodes generic?

I already implemented a highly extendable version in [`experiment-2.ts`](https://github.com/ivanhofer/inlang/blob/lint/source-code/core/src/lint/experiment-2.ts). Generics will not work because you can't have a generic type passed as a generic.

> `withCache` would most likely be a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) object btw that does not "extend" the nodes. The proxy intercepts queries of @inlang/core/query.

I just picked the example you mentioned above.. it could be any other function that adds an additional prop to the nodes.


> Probably not.

Agree. `Proxy`s do not make sense here.

Reply:
Experiment 2 seems the way to go then. Feel free to resolve this discussion